### ISSUE TITLE
0-RTT performes statistically unstable than 1-RTT in HTTP3 Client

### ISSUE THREAD

**Issue by kaiyuhou** [lucas-clemente/quic-go#3150]
**TL;DR:** the `0-RTT` HTTP3 requests are more likely to trigger request retransmissions than `1-RTT` HTTP3 requests. And the `0-RTT` HTTP3 requests have a longer response time than the `1-RTT` HTTP3 request if the `0-RTT` request didn't get a response.

I even don't have any cues to start the investigation. Any comments would be highly appreciated.

#### Experiment Setting

Two machines located in two nearby cities. They maintain a relatively stable and meaningful RTT around `24ms`. 

 - With this setting, we can avoid the weird behavior introduced by too short RTT.

#### Experiment Steps

The client sends `0-RTT`/`1-RTT` HTTP3 request to the server.  

 - Each request repeated `98` times.
 - The `QUIC tokenStore` and `TLS ClientSessionCache` are guaranteed to be valid during `0-RTT` test. I also manually checked it by Wireshark.

#### Expectation

We would expect that all the `0-RTT` request is responded around `24ms`, while all the `1-RTT` request is responded around `48ms`

- If the link is jittery and some packets dropped, the number of longer `0-RTT` requests should similar to the number of longer `1-RTT` requests. 

#### However

 - `0-RTT`
   -  response time `<35ms`: `84`
   -  response time `>35ms`: `14`  (`14` requests don't finish in the `0-RTT` case)
   -  response time `>70ms`: `8` (`8` requests even not finished in the second roundtrip)
 - `1-RTT`
   -  response time `<70ms`: `98` (all requests finished in the second round trip, there are `3` requests in the range of `60~70ms`)

The result shows that the request message carried by the first roundtrip packet is more likely to be ignored by the server. Thus the client has to retransmit the request message. **However, even the retransmit happens, it should perform the same as `1-RTT`, it actually performs worse than `1-RTT`.**

For original stat: [csv](https://github.com/kaiyuhou/quic-go-experiment/blob/main/stat-0rtt-1rtt-100round.csv)
(The [client](https://github.com/kaiyuhou/quic-go-experiment/blob/main/example/kaiyhou_client_quic_single/main.go) and [server](https://github.com/kaiyuhou/quic-go-experiment/blob/main/example/main.go) code can be found here, nothing special though.)

------------

**Why do I notice this** (though not directly related to this problem)

We are trying to use `quic-go` in the data center network and perform some basic measurements. For the HTTP3 request, only the `GET` is `0-RTT` enabled by the `GET0RTT` method. We just cutely thought since a private data center does not need to care about  0RTT replay attach. We can also implement the `POST0RTT` method, and get better performance for the `POST` method. However, because of the problem I mentioned above, the average response time of `POST0RTT` is actually statistically worse than `POST1RTT` in our measurement result. (Please note, the problem I mentioned above is not about the data center network. )

**Comment by marten-seemann** [comment]
There’s actually a very straightforward way to investigate this: quic-go supports qlog. If possible, enable qlog on both client and server side.
The qlogs can be visualized using https://qvis.quictools.info/.

If you want, you can post the qlogs here and I’ll have a look as well.

**Comment by kaiyuhou** [comment]
Thank you for point this out. The qlog visualizer is indeed a powerful tool! 

I re-executed the `0-RTT` experiment with qlog enabled and also captured PCAPs with tls_keylog.   


As listed in [qlog_order_time.csv](https://github.com/kaiyuhou/quic-go-experiment/blob/main/pcaps/2020-04-15-01/qlog_order_time.csv): in those `49` HTTP3 `0-RTT`request
  - there are still `15` requests that didn't get responses in the first roundTrip,
  - while  `11` of the requests didn't get responses in the first two roundTrips.

I pushed all the raw logs. I totally understand that I should do the most preliminary analysis before asking new questions. I will investigate the logs in detail and update my findings tomorrow. (_but if it is due to some obvious reasons, please point it out~_ )

 - [Experiment Folder (containing all the following files)](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01)
   - [Tests](https://github.com/kaiyuhou/quic-go-experiment/blob/main/pcaps/2020-04-15-01/qlog_order_time.csv)
   - [pcap of 50 tests](https://github.com/kaiyuhou/quic-go-experiment/blob/main/pcaps/2020-04-15-01/2020-04-15-01.pcap)
   - [tls_keylog_file.key](https://github.com/kaiyuhou/quic-go-experiment/blob/main/pcaps/2020-04-15-01/tls_keylog_file.key)
   - [clent_qlogs folder](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/client_qlogs)
   - [server_qlogs folder](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/server_qlogs)

**Comment by kaiyuhou** [comment]
I carefully checked these logs with `qvis` and also pcaps. Here is my observation: 

- **the initial 0-RTT data packets (stream 0:0RTT) sending by the client always tend to be ignored by the server (`14/50` cases), while other packets are very rarely to be ignored or dropped.**

I cannot come up with a reason to explain this observation. (I initially thought it may because I triggered these tests too frequently. Some buffers do not have time to take care of themselves. But the problem remains the same after I set the test interval to `1min`.)

Here is an example. 
(The `qvis` shows that the packet is dropped in the middle, but it's not. It's just because the `qvis` cannot distinguish the case of dropping by the network and ignoring by the server. I also captured packets on the server side. The packet arrived at the server.)
![image](https://user-images.githubusercontent.com/11606036/115640171-eb0d7f80-a2db-11eb-9898-4adcca2ae4d2.png)

### Here are more detailed information

I sampled `9` tests for 3 cases and put related qlogs and pcaps into subfolders. Each subfolder only contains logs and pcap related to one http3 request-response test. I also put the full `qvis` sequence screen capture in the folder for easy retrieving:

- **Case 1**: `0-RTT` request responded in `1 roundTrip`: the normal case.
  - [2-0rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/2-0rtt-1893f52581859b62d65cc0) 
  - [4-0rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/4-0rtt-515c1c2484c30ccde6cec0008cc1563af1a496)
  - [43-0rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/43-0rtt-8bb49d5ea542919b13b9)
 - **Case 2**: `0-RTT` request responded in `2 roundTrips`: the initial 0-RTT data packets (stream 0:0RTT) sending by the client are ignored by the server for the reason I don't know yet.
   - [8-1rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/8-1rtt-71da203d3982dcaa256eaeebbda1f6)
   - [27-1rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/27-1rtt-ad5d8c5f82d49eb6172e)
   - [47-1rtt](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/47-1rtt-client_667768e0405531911e195c17e5)
- **Case 3**: `0-RTT` request responded after `2 roundTrips`: the initial 0-RTT data packets (stream 0:0RTT) sending by the client are ignored by the server, then the `appdata PTO timer` was set to some relevantly large value, thus leading to this case.
  - [18-79ms](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/18-79ms-349f01336751d901cc)
  - [25-67ms](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/25-67ms-11ac896056acb7eae69df154a1a7b5c475)
  - [44-90ms](https://github.com/kaiyuhou/quic-go-experiment/tree/main/pcaps/2020-04-15-01/single_roundtrips/44-90ms-b88f9808032777be22)

Now, I feel the real cause of this problem may relate to some cache/session/packet-retrieving design of `quic-go` `server` or `session`. It's a little beyond my knowledge to locate the root cause of this problem.

**Comment by marten-seemann** [comment]
Which version of quic-go are you using? I noticed the `code_version` attribute missing from your qlogs, which seems to imply you're not using v0.20.1: https://github.com/lucas-clemente/quic-go/blob/v0.20.1/qlog/trace.go#L20

**Comment by kaiyuhou** [comment]
I am using the `622ca23d` commit (Jan 31, 2021). It should be the version of v0.19.3. I can try v0.20.1 to see if this problem remains.

**Comment by marten-seemann** [comment]
There were **a lot** of 0-RTT improvements in v0.20.x.

**Comment by kaiyuhou** [comment]
This problem no longer exists after I switched from v0.19.3 to v0.20.1. Thanks for your help!
(At least I proved it's true that there were **a lot** of 0-RTT improvements in v0.20.x 😅)

[RTT of 50 tests](https://github.com/kaiyuhou/quic-go-v201/blob/main/pcaps/2021-04-22/round_qlogs.csv)

**Comment by marten-seemann** [comment]
That’s great to hear. Thanks for reporting!

Can we close this issue?
